ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -806,6 +807,523 @@ class TransactionService(session: LedgerSession) extends LedgerTestSuite(session
+    }
+
+  private[this] val transactionTreeById =
+    LedgerTest("TXTransactionTreeById", "Expose a visible transaction tree by identifier") {
+      context =>
+        for {
+          ledger <- context.participant()
+          party <- ledger.allocateParty()
+          dummy <- ledger.create(party, Dummy(party))
+          tree <- ledger.exercise(party, dummy.exerciseDummyChoice1)
+          byId <- ledger.transactionTreeById(tree.transactionId, party)
+        } yield {
+          assertEquals("The transaction fetched by identifier does not match", tree, byId)
+        }
+    }
+
+  private[this] val invisibleTransactionTreeById =
+    LedgerTest(
+      "TXInvisibleTransactionTreeById",
+      "Do not expose an invisible transaction tree by identifier") { context =>
+      for {
+        Vector(alpha, beta) <- context.participants(2)
+        party <- alpha.allocateParty()
+        intruder <- beta.allocateParty()
+        dummy <- alpha.create(party, Dummy(party))
+        tree <- alpha.exercise(party, dummy.exerciseDummyChoice1)
+        _ <- synchronize(alpha, beta)
+        failure <- beta.transactionTreeById(tree.transactionId, intruder).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.NOT_FOUND, "Transaction not found, or not visible.")
+      }
+    }
+
+  private[this] val transactionTreeByIdNotFound =
+    LedgerTest(
+      "TXTransactionTreeByIdNotFound",
+      "Return NOT_FOUND when looking up an inexistent transaction tree by identifier") { context =>
+      for {
+        ledger <- context.participant()
+        party <- ledger.allocateParty()
+        failure <- ledger.transactionTreeById("a" * 60, party).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.NOT_FOUND, "Transaction not found, or not visible.")
+      }
+    }
+
+  private[this] val transactionTreeByIdNoParty =
+    LedgerTest(
+      "TXTransactionTreeByIdNotFound",
+      "Return INVALID_ARGUMENT when looking up a transaction tree by identifier without specifying a party") {
+      context =>
+        for {
+          ledger <- context.participant()
+          failure <- ledger.transactionTreeById("not-relevant").failed
+        } yield {
+          assertGrpcError(
+            failure,
+            Status.Code.INVALID_ARGUMENT,
+            "Missing field: requesting_parties")
+        }
+    }
+
+  private[this] val transactionTreeByIdSameAsTransactionStream =
+    LedgerTest(
+      "TXTransactionTreeByIdSameAsTransactionStream",
+      "Expose the same events for each transaction as the output of getTransactionTrees") {
+      context =>
+        for {
+          Vector(alpha, beta) <- context.participants(2)
+          submitter <- alpha.allocateParty()
+          listener <- beta.allocateParty()
+          _ <- alpha.create(submitter, AgreementFactory(listener, submitter))
+          _ <- synchronize(alpha, beta)
+          trees <- alpha.transactionTrees(listener, submitter)
+          _ <- eventually {
```

Comment:
I don't think the `eventually` is needed, considering we have a synchronization point before this.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2961/commits/b32f1221504aed4a2bc42e4068e5ee9f8fed8c49